### DISCUSSION TITLE
Img to raw

### DISCUSSION THREAD

**Discussion by PitGR** [micropython/micropython#13261] [ESP32] [answered]
Hello I am PiterGR this is my firt write a discussion. I just want to see what can be done using MicroPython

I am trying to take photos with esp32 cam and then send them to an esp32 and using a st7789 tft screen 1.3" 240x240 . I know that in the st7789py library the blit_buffer method needs a buffer in raw RGB565 format, something like that. Thanks https://gist.github.com/duryding/26a55ee46142d4801a5495b3c88e2859 and https://github.com/shariltumin/esp32-cam-micropython-2022/tree/main.
It is possible to have a buffer of this type WITHOUT using python, just take the photo and send it to the screen 

I hope you can guide me

**Comment by shariltumin** [comment]
These are the settings you can make after ```camera.init()```.

```py
Setting = {
    'pixformat':5,  # 1:RGB565, 2:YUV422, 3:YUV420, 4:GRAYSCALE, 5:COMPRESSED, 
                    # 6:RGB888, 7:RAW, 8:RGB444, 9:RGB555
    'framesize':11, # 1:96x96, 2:160x120, 3:176x144, 4:240x176, 5:240x240
                    # 6:320x240, 7:400x296, 8:480x320, 9:640x480, 10:800x600
                    # 11:1024x768, 12:1280x720, 13:1280x1024, 14:1600x1200
                    # 15:1920x1080, 16:720x1280, 17:864x1536, 18:2048x1536

    'quality':11,   # [0,63] lower number means higher quality
    'contrast':0,   # [-2,2] higher number higher contrast
    'saturation':0, # [-2,2] higher number higher saturation. -2 grayscale
    'brightness':0, # [-2,2] higher number higher brightness. 2 brightest
    'speffect':0,   # 0:,no effect 1:negative, 2:black and white, 3:reddish, 
                    # 4:greenish, 5:blue, 6:retro

    'whitebalance':0, # 0:default, 1:sunny, 2:cloudy, 3:office, 4:home
    'aelevels':0,     # [-2,2] AE Level: Automatic exposure
    'aecvalue':0,     # [0,1200] AEC Value: Automatic exposure control
    'agcgain':0,      # [0,30] AGC Gain: Automatic Gain Control
}
```

In particular, you can try:
```py
camera.pixformat(1) # set format to RGB565
camera.framesize(5) # set framesize to 240x240
```
or
```py
camera.pixformat(4) # set format to GRAYSCALE
camera.framesize(5) # set framesize to 240x240
```

The default values are:
```py
camera.pixformat(5)  # set format to JPG-COMPRESSED
camera.framesize(10) # set framesize to 800x600
```

You can play around with ```camera.quality(x)```, ```camera.contrast(x)```, and other settings for your camera, display and environment to get the best results.

I got good results writing bytes from camptured image directly to ST7735.

Part of my test code (the main loop):

```py
def display():
    w,h=160-1,80-1 # screen image size
    s0=160*20*2    # start at img[160*20*2] i.e 20 row down
    sx=160*100*2   # end at img[160*100*2] i.e 100-20=80 rows
    loc=tft._setwindowloc
    take=camera.capture
    put=tft._writedata
    while True:
       img=take()
       loc((0,0),(w,h)) # top left, bottom right
       put(img[s0:sx])  # send portion of img to TFT screen
```

I haven't tried with the ST7789, but if you can get "_setwindowloc" and "_writedata" or equivalent to work on the ST7789, then the rest is easy.

Best of luck. Merry Christmas and Happy New Year 2024.

**Reply by PitGR** [reply]
Thanks shariltumin 

For you I do my P!tCarSpy used firmwares-20230717 thanks you. I have fun while learn MicroPython, JavaScript, Python and AppInventor,, my car can be controlled with them. Using smartphone or PC

Now I'm trying to see the photos on a 240x240 SPI tft screen 

I know the settings, I have tried them several times but it happened that 

import camera

def capture():
  buf = camera.capture()
  if buf:
    print(len(buf))
    f = open('img.raw','wb')
    f.write(buf)
    f.close()

 camera.deinit()

camera.init()
camera.pixformat(1) # set format to RGB565
camera.framesize(5) # set framesize to 240x240

capture()

buf is None, not bytes ;(

st7789 library, blit_buffer method
self._set_window(x, y, x + width - 1, y + height - 1)  # column address and row address
self._write(None, buffer) # buffer I think that RGB565 
is the same 
loc=tft._setwindowloc
put=tft._writedata

used this utils/img2rgb565.py 
but I dont want to use. it is possible to have the buffer only using Micropython?  I'm sorry, I don't know about many things.

Could you explain to me why those values, 160, 20, 100, 2 what is the meaning 
s0=160*20*2    # start at img[160*20*2] i.e 20 row down
sx=160*100*2   # end at img[160*100*2] i.e 100-20=80 rows

thank you again shariltumin for your great contribution

PiterGR sends you a hug from Mexico and happy holidays

**Comment by shariltumin** [comment]
Hi,

With these code:
```py
camera.init()
camera.pixformat(1) # set format to RGB565
camera.framesize(5) # set framesize to 240x240
buf = camera.capture()
```

You will get an RGB565 formatted image in buf which you can simply write to ST7789 WITHOUT using any conversion program such as utils/img2rgb565.py because the data in buf is already in RGB565 format.

Don't worry about my s0 and sx, they are there because I need to trim the image to fit my small display. In your case they will be ```_set_window(0, 0, 239, 239)``` and ```._write(None, buf)```.

Please try with a simple program first; 
1. init the camera
2. capture an image 
3. send to display

I hope this will work for you.

**Reply by PitGR** [reply]
I tried but the result is not what I expected

I saw buffer ttps://github.com/rdagger/micropython-ili9341/blob/master/images/MicroPython128x128.raw 
f = open('MicroPython128x128.raw','rb')
buf= f.read()
len(buf); 32768
buf[0:10]; b'\x00\x00\x00\x00\x00\x00\x00\x00\x00\x00' 
all buf bytes or hex 

then try to find a similarity but 

import camera
camera.init()

camera.pixformat(1)
buf = camera.capture()
len(buf); 20105
buf[0:10]; b'\xff\xd8\xff\xe0\x00\x10JFIF'

camera.pixformat(4)
buf = camera.capture()
len(buf); 19763
buf[0:10]; b'\xff\xd8\xff\xe0\x00\x10JFIF'

camera.pixformat(5)
buf = camera.capture()
len(buf); 16304
buf[0:10]; b'\xff\xd8\xff\xe0\x00\x10JFIF'

camera.pixformat(8)
buf = camera.capture()
len(buf); 20511
buf[0:10]; b'\xff\xd8\xff\xe0\x00\x10JFIF'

camera.pixformat(9)
buf = camera.capture()
len(buf); 21306
buf[0:10]; b'\xff\xd8\xff\xe0\x00\x10JFIF'

Apparently all the pixformats are in JPEG or something like that. I'm confused.

also try. I make tests configuring pixformat and framesize it fails
camera.pixformat(1) # set format to RGB565
camera.framesize(5) # set framesize to 240x240
buf = camera.capture()
[0;31mE (102790) camera: Camera Capture Failed[0m
UpyCraft and Thonny doesn't show anything

I'm also checking capture_bmp. but I find myself lost 

Thanks, I'll look for other ideas

**Comment by shariltumin** [comment, marked as answer]
Hi,

Thanks for testing. Yes, the firmware has some bugs. The method I wrote about was for pre-2022 firmwares. I had released a new version [here](https://github.com/shariltumin/esp32-cam-micropython-2022/tree/main/Christmas-2023-Edition).

Please give it a try. The sample scripts are for ST7735 only. Hopefully it will work better this time.

**Reply by PitGR** [reply]
Thanks and sorry I don't see the message early but this version is very well.

I try this but it is slow 

import machine
import st7789py as st7789
import io
import camera

camera.init()
camera.framesize(5) # set framesize to 240x240
buf = camera.capture_bmp()

spi = machine.SPI(1, baudrate=4000000, polarity=1, phase=0, firstbit=machine.SPI.MSB)
display = st7789.ST7789(
  spi, 240, 240,
  xstart=0, ystart=0,
  reset=machine.Pin(2, machine.Pin.OUT),
  dc=machine.Pin(15, machine.Pin.OUT),
)

if buf_stream.read(2) == b'BM':  # header
  dummy = buf_stream.read(8)  # file size(4), creator bytes(4)
  offset = int.from_bytes(buf_stream.read(4), 'little')
  hdrsize = int.from_bytes(buf_stream.read(4), 'little')
  width = int.from_bytes(buf_stream.read(4), 'little')
  height = int.from_bytes(buf_stream.read(4), 'little')
  height = -width
  if int.from_bytes(buf_stream.read(2), 'little') == 1:  # planes must be 1
    depth = int.from_bytes(buf_stream.read(2), 'little')
    if depth == 24 and int.from_bytes(buf_stream.read(4), 'little') == 0:  # compress method == uncompressed
      print("Image size:", width, "x", height)
      rowsize = (width * 3 + 3) & ~3
      if height < 0:
        height = -height
        flip = False
      else:
        flip = True
      w, h = width, height
      if w > width:
        w = width
      if h > 240:
        h = 240
      for row in range(h):
        if flip:
          pos = offset + (height - 1 - row) * rowsize
        else:
          pos = offset + row * rowsize
        if buf_stream.tell() != pos:
          dummy = buf_stream.seek(pos)
        for col in range(w):
          bgr = buf_stream.read(3) # Blue Green Red 
          rgb = st7789.color565(bgr[2], bgr[1], bgr[0])
          display.pixel(col, row, rgb)

The next year, tomorrow I will start to make a pcb 

Again thanks for amazing help. I hope a day to become as good as you but step by step.

Happy New Year, 2024. I see soon a strong hug
